File: leetcode/7.reverse-integer.cpp
class Solution {
public:
    int reverse(int x) {
        int res = 0;
        while (x != 0) {
            int pop = x % 10; // gets last number
            x /= 10;
            if (res > INT_MAX / 10 || (res == INT_MAX / 10 && pop > 7)) return 0;
            if (res < INT_MIN / 10 || (res == INT_MIN / 10 && pop < -8)) return 0;
            res = res * 10 + pop;
        }
        
        return res;
    }
};

class Solution2 {
public:
    int reverse(int x) {
        
        int res = 0;
        while(x != 0)
        {
            int pop = x % 10;
            x = x / 10;

            int candidate = res * 10 + pop;

            if (candidate / 10 != res)
            {
                return 0;
            }

            res = candidate;            
        }

        return res;
    }
};

File: leetcode/7.reverse-integer.py
class Solution:
    def reverse(self, x: int) -> int:
 
        max_integer = 2 ** 31 - 1
        min_integer = - 2 ** 31

        sign = 1 if x >= 0 else -1
        x = abs(x)
        reverse = 0
        
        while x > 0:
            reverse = reverse * 10 + x % 10
            x = x // 10
            
        res = sign * reverse
        return res if res >= min_integer and res <= max_integer else 0